src/backend.rs

```diff
@@ -205,88 +346,639 @@ impl StackDepth {
     }
 }
 
-pub fn current_stack_depth(ctx: &Context) -> StackDepth {
-    ctx.sp_depth
-}
+fn expand_stack(ctx: &mut Context, by: u32) {
+    use std::iter;
 
-pub fn restore_stack_depth(ctx: &mut Context, stack_depth: StackDepth) {
-    ctx.sp_depth = stack_depth;
-}
+    if by == 0 {
+        return;
+    }
+
+    let new_stack_size = (ctx.block_state.stack_map.len() + by as usize).next_power_of_two();
+    let additional_elements = new_stack_size - ctx.block_state.stack_map.len();
+    ctx.block_state
+        .stack_map
+        .extend(iter::repeat(false).take(additional_elements));
 
-fn push_i32(ctx: &mut Context, gpr: GPR) {
-    // For now, do an actual push (and pop below). In the future, we could
-    // do on-the-fly register allocation here.
-    ctx.sp_depth.reserve(1);
     dynasm!(ctx.asm
-        ; push Rq(gpr)
+        ; sub rsp, additional_elements as i32
     );
-    ctx.regs.release_scratch_gpr(gpr);
 }
 
-fn pop_i32(ctx: &mut Context) -> GPR {
-    ctx.sp_depth.free(1);
-    let gpr = ctx.regs.take_scratch_gpr();
-    dynasm!(ctx.asm
-        ; pop Rq(gpr)
+// TODO: Make this generic over `Vec` or `ArrayVec`?
+fn stack_slots(ctx: &mut Context, count: u32) -> Vec<i32> {
```

Comment:
Is this function used anywhere?

Reply:
No, it's unfinished work to implement stack allocation (that doesn't emit `push`/`pop`). There are quite a few optimisations that AFAICT cannot be implemented in linear time without this (for example, you can see that we do a linear scan over the stack on `set_local`).